lib/grandpa/grandpa.go

```diff
@@ -449,188 +424,300 @@ func (s *Service) primaryBroadcastCommitMessage() {
 // playGrandpaRound executes a round of GRANDPA
 // at the end of this round, a block will be finalised.
 func (s *Service) playGrandpaRound() error {
-	logger.Debugf("starting round %d with set id %d",
-		s.state.round, s.state.setID)
+	currentRound := s.state.round
+	currentSetID := s.state.setID
+
 	start := time.Now()
 
-	ctx, cancel := context.WithCancel(s.ctx)
-	defer cancel()
+	logger.Debugf("starting round %d with set id %d",
+		currentRound, currentSetID)
 
-	isPrimary, err := s.handleIsPrimary()
-	if err != nil {
-		return err
-	}
+	cancel := make(chan struct{})
+	defer close(cancel)
```

Comment:
This should be closed if the service context is canceled. If we use a context inside the Service, we might want to just pass in `s.ctx` to `s.finalizationEngine`.

Transforming the context to a signal channel would require an extra goroutine which is meh.

Alternatively you could change `s.ctx` to be a stop signal channel, but I'm not sure about the implications.